File: authority-contract/src/methods/external_key_authority.cpp
/* Copyright 2026 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <string>
#include <vector>
#include <stddef.h>
#include <stdint.h>

#include "Cryptography.h"
#include "KeyValue.h"
#include "Environment.h"
#include "Message.h"
#include "Response.h"
#include "Types.h"
#include "Util.h"
#include "Value.h"
#include "WasmExtensions.h"

#include "contract/base.h"
#include "identity/identity.h"
#include "identity/policy_agent.h"          // trusted-issuer map + verify_credential
#include "identity/signature_authority.h"
#include "identity/common/Credential.h"
#include "identity/common/SigningContext.h"
#include "identity/common/SigningContextManager.h"
#include "identity/crypto/RSAPublicKey.h"   // RSA public key, for the session key's signature
#include "authority/wallet_key_authority.h" // WALLET_VERIFYING_KEY_CREDENTIAL_TYPE
#include "authority/external_key_authority.h"

// -----------------------------------------------------------------
// METHOD: initialize_contract
//   Creation hook. Sets up the inherited identity signing contexts + VC store,
//   an empty trusted-issuers map (so a wallet_key_authority can be registered as
//   a trusted issuer of WalletVerifyingKeyCredentials), and a dedicated signing
//   context that this authority signs the credentials it issues with. It holds NO
//   ledger key: it delegates the ledger check to the wallet_key_authority that
//   issued the WalletVerifyingKeyCredential it consumes.
// -----------------------------------------------------------------
bool ww::authority::external_key_authority::initialize_contract(const Environment& env)
{
    if (! ww::identity::identity::initialize_contract(env))
        return false;

    if (! ww::identity::policy_agent::initialize_trusted_issuers())
        return false;

    std::vector<std::string> issuer_path = {EXTERNAL_KEY_AUTHORITY_ISSUER_CONTEXT};
    ww::identity::SigningContextManager manager = ww::identity::identity::get_context_manager();
    if (! manager.add_context(false, "external key authority issuer context", issuer_path))
        return false;

    return true;
}

// -----------------------------------------------------------------
// Verify a base64 signature over the payload string using a PEM public key.
//
// The wallet signs with its contract signing key (the WalletVerifyingKeyCredential's
// "verifying_key" claim is its public half), so its signature is checked with the
// wawaka ecdsa primitive -- the same SHA-256 ECDSA scheme PDO contract keys sign
// with. The session key signs with its rsa key (RSASSA-PKCS1-v1_5 / SHA-256). The
// message is the payload string bytes.
// -----------------------------------------------------------------
static bool verify_ec_signature(
    const std::string& payload, const std::string& b64_signature, const std::string& key)
{
    ww::types::ByteArray message(payload.begin(), payload.end());
    ww::types::ByteArray signature;
    if (! ww::crypto::b64_decode(b64_signature, signature))
        return false;

    return ww::crypto::ecdsa::verify_signature(message, key, signature);
}

static bool verify_rsa_signature(
    const std::string& payload, const std::string& b64_signature, const std::string& key)
{
    ww::types::ByteArray message(payload.begin(), payload.end());
    ww::types::ByteArray signature;
    if (! ww::crypto::b64_decode(b64_signature, signature))
        return false;

    pdo_contracts::crypto::signing::RSAPublicKey public_key;
    if (! public_key.Deserialize(key))
        return false;
    return public_key.VerifySignature(message, signature);
}

// -----------------------------------------------------------------
// Build and sign a publicKeyCredential asserting that the wallet (subject) controls
// the session key (the "key" claim). Signed with this authority's issuer signing
// context, exactly as signature_authority::sign_credential signs a caller-supplied
// credential.
// -----------------------------------------------------------------
static bool build_public_key_credential(
    const Environment& env,
    const std::string& wallet_did,
    const std::string& session_key,
    ww::value::Object& serialized_vc_out,
    std::string& error_msg)
{
    ww::identity::Credential credential;
    credential.type_ = {PUBLIC_KEY_CREDENTIAL_TYPE};
    credential.issuer_.id_ = env.contract_id_;
    credential.credentialSubject_.subject_.id_ = wallet_did;
    credential.credentialSubject_.claims_.set_string("key", session_key.c_str());

    ww::value::Object credential_object;
    if (! credential.serialize(credential_object))
    {
        error_msg.assign("unexpected error, failed to serialize the credential");
        return false;
    }

    const std::vector<std::string> context_path = {EXTERNAL_KEY_AUTHORITY_ISSUER_CONTEXT};
    const ww::identity::IdentityKey identity(env.contract_id_, context_path);
    ww::identity::SigningContextManager manager = ww::identity::identity::get_context_manager();

    ww::identity::SigningContext context;
    std::vector<std::string> extended_path;
    if (! manager.find_context(context_path, extended_path, context))
    {
        error_msg.assign("unexpected error, failed to locate the issuer signing context");
        return false;
    }
    context.set_context_path(extended_path);

    ww::identity::VerifiableCredential vc;
    if (! vc.build(credential_object, identity, context))
    {
        error_msg.assign("unexpected error, failed to sign the credential");
        return false;
    }

    if (! vc.serialize(serialized_vc_out))
    {
        error_msg.assign("unexpected error, failed to serialize the signed credential");
        return false;
    }

    return true;
}

// -----------------------------------------------------------------
// METHOD: sign_credential
//   This authority's customization of signature_authority::sign_credential. It
//   binds an external session key to a wallet from a WalletVerifyingKeyCredential the
//   consumer got from a trusted wallet_key_authority:
//     - verify that credential against a registered trusted issuer, and take the
//       wallet's DID (its subject) and verifying key (its "verifying_key" claim);
//     - confirm the wallet and the session key signed the same payload -- the
//       wallet against its verifying key, the session key against the public half
//       the payload carries -- and that the payload names this wallet;
//     - issue a signed publicKeyCredential (subject = wallet DID, "key" = session key).
//
// JSON PARAMETERS:
//   EXTERNAL_KEY_AUTHORITY_SIGN_CREDENTIAL_PARAM_SCHEMA
//
// RETURNS:
//   VERIFIABLE_CREDENTIAL_SCHEMA -- the signed publicKeyCredential
// -----------------------------------------------------------------
bool ww::authority::external_key_authority::sign_credential(
    const Message& msg, const Environment& env, Response& rsp)
{
    ASSERT_INITIALIZED(rsp);

    ASSERT_SUCCESS(rsp, msg.validate_schema(EXTERNAL_KEY_AUTHORITY_SIGN_CREDENTIAL_PARAM_SCHEMA),
                   "invalid request, missing required parameters");

    // ---------- verify the WalletVerifyingKeyCredential against a trusted issuer ----
    ww::value::Object wskc_object;
    ASSERT_SUCCESS(rsp, msg.get_value("wallet_verifying_key_credential", wskc_object),
                   "invalid request, missing wallet_verifying_key_credential");

    ww::identity::VerifiableCredential wskc;
    ASSERT_SUCCESS(rsp, wskc.deserialize(wskc_object),
                   "invalid request, ill-formed wallet verifying key credential");
    ASSERT_SUCCESS(rsp, ! wskc.credential_.type_.empty() &&
                            wskc.credential_.type_[0] == WALLET_VERIFYING_KEY_CREDENTIAL_TYPE,
                   "invalid request, credential is not a WalletVerifyingKeyCredential");

    // verified against the trusted wallet_key_authority registered for this type
    ASSERT_SUCCESS(rsp, ww::identity::policy_agent::verify_credential(
                            wskc_object, wskc, WALLET_VERIFYING_KEY_CREDENTIAL_TYPE),
                   "invalid request, wallet verifying key credential is not from a trusted issuer");

    // the wallet's DID is the credential subject; its verifying key is the claim
    const std::string wallet_did(wskc.credential_.credentialSubject_.subject_.id_);
    const std::string verifying_key(wskc.credential_.credentialSubject_.claims_.get_string("verifying_key"));
    ASSERT_SUCCESS(rsp, ! verifying_key.empty(),
                   "invalid request, wallet verifying key credential has no verifying_key claim");

    // ---------- the wallet and the session key must have signed the same payload -
    const std::string wallet_payload(msg.get_string("wallet_attestation.payload"));
    const std::string wallet_signature(msg.get_string("wallet_attestation.signature"));
    const std::string session_payload(msg.get_string("session_key_attestation.payload"));
    const std::string session_signature(msg.get_string("session_key_attestation.signature"));

    ASSERT_SUCCESS(rsp, wallet_payload == session_payload,
                   "invalid request, the wallet and the session key signed different payloads");

    // the payload names the wallet and carries the session key's public half
    ww::value::Object payload;
    ASSERT_SUCCESS(rsp, payload.deserialize(wallet_payload.c_str()) &&
                            payload.validate_schema(EXTERNAL_KEY_PAYLOAD_SCHEMA),
                   "invalid request, ill-formed binding payload");
    const std::string identity(payload.get_string("identity"));
    const std::string session_key(payload.get_string("session_key"));

    ASSERT_SUCCESS(rsp, identity == wallet_did,
                   "invalid request, the payload identity does not match the wallet credential subject");

    // ---------- verify both signatures over the payload ----------
    // the wallet, using its verifying key from the credential; the session key,
    // proving possession
    ASSERT_SUCCESS(rsp, verify_ec_signature(wallet_payload, wallet_signature, verifying_key),
                   "invalid request, the wallet signature could not be verified");
    ASSERT_SUCCESS(rsp, verify_rsa_signature(session_payload, session_signature, session_key),
                   "invalid request, the session key signature could not be verified");

    // ---------- issue the publicKeyCredential ----------
    std::string error_msg;
    ww::value::Object serialized_vc_out;
    ASSERT_SUCCESS(rsp, build_public_key_credential(env, wallet_did, session_key, serialized_vc_out, error_msg),
                   error_msg.c_str());

    return rsp.value(serialized_vc_out, false);
}

File: authority-contract/pdo/authority/plugins/external_key_authority.py
# Copyright 2026 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
import logging

from pdo.contract import invocation_request
from pdo.submitter.create import create_submitter

import pdo.client.builder as pbuilder
import pdo.client.builder.command as pcommand
import pdo.client.builder.contract as pcontract
import pdo.client.builder.shell as pshell
import pdo.client.commands.contract as pcontract_cmd
import pdo.common.crypto as pcrypto

import pdo.client.plugins.common as common
import pdo.identity.plugins.signature_authority as signature_authority
import pdo.identity.plugins.policy_agent as policy_agent
import pdo.authority.plugins.wallet_key_authority as wallet_key_authority
import pdo.authority.session_key as session_key

logger = logging.getLogger(__name__)

WALLET_VERIFYING_KEY_CREDENTIAL_TYPE = "WalletVerifyingKeyCredential"
PDO_DID_PREFIX = "did:pdo:"

__all__ = [
    'op_initialize',
    'op_get_verifying_key',
    'op_get_extended_verifying_key',
    'op_register_signing_context',
    'op_describe_signing_context',
    'op_list_signing_contexts',
    'op_add_vc',
    'op_get_vc_list',
    'op_get_vp',
    'op_sign_credential',
    'op_verify_credential',
    'op_register_trusted_issuer',
    'op_list_trusted_issuers',
    'cmd_register_signing_context',
    'cmd_list_signing_contexts',
    'cmd_get_verifying_key',
    'cmd_verify_credential',
    'cmd_register_trusted_issuer',
    'cmd_list_trusted_issuers',
    'cmd_create_external_key_authority',
    'cmd_sign_credential',
    'cmd_bind_external_key',
    'do_external_key_authority',
    'do_external_key_authority_contract',
    'load_commands',
]

# -----------------------------------------------------------------
# external_key_authority inherits its identity operations from signature_authority
# and its trusted-issuer operations from policy_agent (used to trust a
# wallet_key_authority). sign_credential is customized (see op_sign_credential):
# it binds an external session key to a wallet from a trusted WalletVerifyingKeyCredential.
# -----------------------------------------------------------------
op_initialize = signature_authority.op_initialize
op_get_verifying_key = signature_authority.op_get_verifying_key
op_get_extended_verifying_key = signature_authority.op_get_extended_verifying_key
op_register_signing_context = signature_authority.op_register_signing_context
op_describe_signing_context = signature_authority.op_describe_signing_context
op_list_signing_contexts = signature_authority.op_list_signing_contexts
op_add_vc = signature_authority.op_add_vc
op_get_vc_list = signature_authority.op_get_vc_list
op_get_vp = signature_authority.op_get_vp
op_sign_with_contract_key = signature_authority.op_sign_with_contract_key
op_verify_credential = signature_authority.op_verify_credential

op_get_contract_metadata = common.op_get_contract_metadata
op_register_trusted_issuer = policy_agent.op_register_trusted_issuer
op_list_trusted_issuers = policy_agent.op_list_trusted_issuers

cmd_register_signing_context = signature_authority.cmd_register_signing_context
cmd_list_signing_contexts = signature_authority.cmd_list_signing_contexts
cmd_get_verifying_key = signature_authority.cmd_get_verifying_key
cmd_verify_credential = signature_authority.cmd_verify_credential
cmd_register_trusted_issuer = policy_agent.cmd_register_trusted_issuer
cmd_list_trusted_issuers = policy_agent.cmd_list_trusted_issuers


# -----------------------------------------------------------------
# -----------------------------------------------------------------
class op_sign_credential(pcontract.contract_op_base) :
    name = "sign_credential"
    help = "bind an external session key to a wallet and issue a publicKeyCredential for it"

    @classmethod
    def add_arguments(cls, subparser) :
        subparser.add_argument(
            '--wallet-verifying-key-credential',
            help='the WalletVerifyingKeyCredential from a trusted wallet_key_authority',
            type=pbuilder.invocation_parameter, required=True)
        subparser.add_argument(
            '--wallet-attestation',
            help='the binding payload signed by the wallet ({payload, signature})',
            type=pbuilder.invocation_parameter, required=True)
        subparser.add_argument(
            '--session-key-attestation',
            help='the same binding payload signed by the session key ({payload, signature})',
            type=pbuilder.invocation_parameter, required=True)

    @classmethod
    def invoke(cls, state, session_params, wallet_verifying_key_credential, wallet_attestation,
               session_key_attestation, **kwargs) :
        session_params['commit'] = False

        message = invocation_request(
            'sign_credential',
            wallet_verifying_key_credential=wallet_verifying_key_credential,
            wallet_attestation=wallet_attestation,
            session_key_attestation=session_key_attestation)
        result = pcontract_cmd.send_to_contract(state, message, **session_params)
        cls.log_invocation(message, result)

        return result


# -----------------------------------------------------------------
# -----------------------------------------------------------------
class cmd_create_external_key_authority(pcommand.contract_command_base) :
    name = "create"
    help = "create an external key authority"

    @classmethod
    def add_arguments(cls, subparser) :
        subparser.add_argument('-c', '--contract-class', help='Name of the contract class', type=str)
        subparser.add_argument('-e', '--eservice-group', help='Name of the enclave service group to use', type=str)
        subparser.add_argument('-f', '--save-file', help='File where contract data is stored', type=str)
        subparser.add_argument('-p', '--pservice-group', help='Name of the provisioning service group to use', type=str)
        subparser.add_argument('-r', '--sservice-group', help='Name of the storage service group to use', type=str)
        subparser.add_argument('--source', help='File that contains contract source code', type=str)
        subparser.add_argument('--extra', help='Extra data associated with the contract file', nargs=2, action='append')

    @classmethod
    def invoke(cls, state, context, **kwargs) :
        save_file = pcontract_cmd.get_contract_from_context(state, context)
        if save_file :
            return save_file

        # create the contract
        save_file = pcontract_cmd.create_contract_from_context(state, context, 'external_key_authority', **kwargs)
        context['save_file'] = save_file

        session = pbuilder.SessionParameters(save_file=save_file)
        pcontract.invoke_contract_op(
            op_initialize,
            state, context, session,
            context['description'],
            **kwargs)

        # create the wallet key authority that issues the credentials this authority
        # consumes, unless it already exists
        wka_context = context.get_context('wallet_key_authority_context')
        wka_save_file = pcontract_cmd.get_contract_from_context(state, wka_context)
        if not wka_save_file :
            wka_save_file = pcommand.invoke_contract_cmd(
                wallet_key_authority.cmd_create_wallet_key_authority,
                state, wka_context,
                **kwargs)

        # trust the wallet key authority as an issuer of WalletVerifyingKeyCredentials
        pcommand.invoke_contract_cmd(
            cmd_register_trusted_issuer,
            state, context,
            issuer=wka_context.path,
            path=['wallet_key_authority'],
            credential_types=[WALLET_VERIFYING_KEY_CREDENTIAL_TYPE],
            **kwargs)

        cls.display('created external key authority in {}'.format(save_file))
        return save_file


# -----------------------------------------------------------------
# -----------------------------------------------------------------
class cmd_sign_credential(pcommand.contract_command_base) :
    name = "sign_credential"
    help = "bind an external session key to a wallet and issue a publicKeyCredential for the session key"

    @classmethod
    def add_arguments(cls, subparser) :
        subparser.add_argument(
            '--wallet-verifying-key-credential',
            help='file with the WalletVerifyingKeyCredential from a trusted wallet_key_authority',
            type=str, required=True)
        subparser.add_argument(
            '--wallet-attestation',
            help='file with the binding payload signed by the wallet ({payload, signature})',
            type=str, required=True)
        subparser.add_argument(
            '--session-key-attestation',
            help='file with the same payload signed by the session key ({payload, signature})',
            type=str, required=True)
        subparser.add_argument(
            '--credential',
            help='file where the issued verifiable credential will be written',
            type=str, required=True)

    @classmethod
    def invoke(cls, state, context, wallet_verifying_key_credential, wallet_attestation,
               session_key_attestation, credential, **kwargs) :
        save_file = pcontract_cmd.get_contract_from_context(state, context)
        if not save_file :
            raise ValueError('external key authority contract must be created and initialized')

        with open(wallet_verifying_key_credential, "r") as fp :
            wskc = json.load(fp)
        with open(wallet_attestation, "r") as fp :
            wallet_attestation_data = json.load(fp)
        with open(session_key_attestation, "r") as fp :
            session_key_attestation_data = json.load(fp)

        session = pbuilder.SessionParameters(save_file=save_file)
        vc = pcontract.invoke_contract_op(
            op_sign_credential,
            state, context, session,
            wallet_verifying_key_credential=wskc,
            wallet_attestation=wallet_attestation_data,
            session_key_attestation=session_key_attestation_data,
            **kwargs)

        with open(credential, "w") as fp :
            fp.write(vc)

        cls.display('saved publicKeyCredential to {}'.format(credential))
        return True


## -----------------------------------------------------------------
## bind an external key to a wallet
## -----------------------------------------------------------------
class cmd_bind_external_key(pcommand.contract_command_base) :
    """Generate an external RSA key and bind it to a wallet
    The wallet must be created and the wallet key authority that issues its
    verifying key credential must be trusted by this external key authority.
    """

    name = "bind_external_key"
    help = "generate an external RSA key and bind it to a wallet"

    @classmethod
    def add_arguments(cls, subparser) :
        subparser.add_argument(
            '-w', '--wallet',
            help='context of the wallet to bind the external key to',
            type=str, required=True)
        subparser.add_argument(
            '--keys-dir',
            help='directory where the generated external RSA key pair is written',
            type=str, required=True)

    @classmethod
    def get_wallet_credential(cls, state, context, wallet_context, wallet_session, wallet_contract, **kwargs) :
        """Get a WalletVerifyingKeyCredential for the wallet and store it in the wallet

        The credential comes from the wallet key authority wired into this external
        key authority's context as the wallet_key_authority_context.
        """

        wka_context = context.get_context('wallet_key_authority_context')
        wka_save_file = pcontract_cmd.get_contract_from_context(state, wka_context)
        if wka_save_file is None :
            # A caller may have built a fresh context tree that never carried
            # the wka's contract_id into wallet_key_authority_context (e.g.
            # one generated per call, disconnected from the tree used at
            # creation time). Recover it from this eka's own trusted-issuer
            # list, which always trusts exactly one issuer for
            # WalletVerifyingKeyCredential (registered by
            # cmd_create_external_key_authority).
            save_file = pcontract_cmd.get_contract_from_context(state, context)
            session = pbuilder.SessionParameters(save_file=save_file)
            raw_issuers = pcontract.invoke_contract_op(
                op_list_trusted_issuers, state, context, session, **kwargs)
            issuers = json.loads(raw_issuers) if isinstance(raw_issuers, str) else (raw_issuers or {})
            wka_id = next(
                (issuer_id for issuer_id, entries in issuers.items()
                 for entry in (entries or [])
                 if WALLET_VERIFYING_KEY_CREDENTIAL_TYPE in (entry.get('credential_types') or [])),
                None)
            if wka_id is None :
                raise ValueError('external key authority has no trusted wallet_key_authority registered')
            wka_context.set('contract_id', wka_id)
            wka_save_file = pcontract_cmd.get_contract_from_context(state, wka_context)
            if wka_save_file is None :
                raise ValueError('unable to locate wallet key authority contract on the ledger')
        wka_session = pbuilder.SessionParameters(save_file=wka_save_file)

        # the authority attests the wallet's verifying key from its ledger attestation
        # and metadata, so we collect both from the ledger and the wallet contract
        ledger_submitter = create_submitter(state.get(['Ledger']))
        ledger_attestation = ledger_submitter.get_contract_info(wallet_contract.contract_id)

        contract_metadata = pcontract.invoke_contract_op(
            op_get_contract_metadata,
            state, wallet_context, wallet_session,
            **kwargs)
        contract_metadata = json.loads(contract_metadata)

        wallet_credential = pcontract.invoke_contract_op(
            wallet_key_authority.op_sign_credential,
            state, wka_context, wka_session,
            wallet_contract.contract_id,
            wallet_contract.creator_id,
            ledger_attestation,
            contract_metadata,
            **kwargs)
        wallet_credential = json.loads(wallet_credential)

        # store the credential in the wallet so a later bind can reuse it
        pcontract.invoke_contract_op(
            op_add_vc,
            state, wallet_context, wallet_session,
            wallet_credential,
            **kwargs)

        return wallet_credential

    @classmethod
    def invoke(cls, state, context, wallet, keys_dir, **kwargs) :
        save_file = pcontract_cmd.get_contract_from_context(state, context)
        if not save_file :
            raise ValueError('external key authority contract must be created and initialized')
        session = pbuilder.SessionParameters(save_file=save_file)

        # resolve the wallet from its context; its contract id is the DID that the
        # binding payload and the issued credentials are all anchored to
        wallet_context = pbuilder.Context(state, wallet)
        wallet_save_file = pcontract_cmd.get_contract_from_context(state, wallet_context)
        wallet_contract = pcontract_cmd.get_contract(state, wallet_save_file)
        wallet_did = PDO_DID_PREFIX + wallet_contract.contract_id
        wallet_session = pbuilder.SessionParameters(save_file=wallet_save_file)

        # generate the external key pair that will be bound to the wallet
        session_key.generate_rsa_keypair(keys_dir)
        session_public_pem = session_key.load_public_pem(keys_dir)

        # the binding needs a WalletVerifyingKeyCredential; reuse the one the wallet
        # already holds, otherwise get a fresh one from the trusted authority
        vc_map = json.loads(pcontract.invoke_contract_op(
            op_get_vc_list,
            state, wallet_context, wallet_session,
            **kwargs))
        wallet_credential = vc_map.get(WALLET_VERIFYING_KEY_CREDENTIAL_TYPE)
        if wallet_credential is None :
            wallet_credential = cls.get_wallet_credential(
                state, context,
                wallet_context, wallet_session, wallet_contract,
                **kwargs)

        # the wallet and the session key sign the same binding payload; the wallet
        # signs with its contract key so the signature verifies against its ledger key
        payload = session_key.build_payload(wallet_did, session_public_pem)
        b64_payload = pcrypto.byte_array_to_base64(pcrypto.string_to_byte_array(payload))
        wallet_signature = pcontract.invoke_contract_op(
            op_sign_with_contract_key,
            state, wallet_context, wallet_session,
            b64_payload,
            **kwargs)
        wallet_attestation = { 'payload' : payload, 'signature' : json.loads(wallet_signature) }
        session_key_attestation = { 'payload' : payload, 'signature' : session_key.sign_payload_b64(keys_dir, payload) }

        # the authority checks the credential and both signatures and issues a
        # publicKeyCredential for the session key
        public_key_credential = pcontract.invoke_contract_op(
            op_sign_credential,
            state, context, session,
            wallet_credential,
            wallet_attestation,
            session_key_attestation,
            **kwargs)
        public_key_credential = json.loads(public_key_credential)

        # store the issued credential in the wallet
        pcontract.invoke_contract_op(
            op_add_vc,
            state, wallet_context, wallet_session,
            public_key_credential,
            **kwargs)

        cls.display('bound external key to wallet {}'.format(wallet_contract.contract_id))
        return public_key_credential


# -----------------------------------------------------------------
# Create the generic, shell independent version of the aggregate command
# -----------------------------------------------------------------
__operations__ = [
    op_initialize,
    op_get_verifying_key,
    op_get_extended_verifying_key,
    op_register_signing_context,
    op_describe_signing_context,
    op_list_signing_contexts,
    op_add_vc,
    op_get_vc_list,
    op_get_vp,
    op_sign_credential,
    op_verify_credential,
    op_register_trusted_issuer,
    op_list_trusted_issuers,
]

do_external_key_authority_contract = pcontract.create_shell_command('external_key_authority_contract', __operations__)

__commands__ = [
    cmd_get_verifying_key,
    cmd_register_signing_context,
    cmd_list_signing_contexts,
    cmd_verify_credential,
    cmd_register_trusted_issuer,
    cmd_list_trusted_issuers,
    cmd_create_external_key_authority,
    cmd_sign_credential,
    cmd_bind_external_key,
]

do_external_key_authority = pcommand.create_shell_command('external_key_authority', __commands__)


# -----------------------------------------------------------------
# Enable binding of the shell independent version to a pdo-shell command
# -----------------------------------------------------------------
def load_commands(cmdclass) :
    pshell.bind_shell_command(cmdclass, 'external_key_authority', do_external_key_authority)
    pshell.bind_shell_command(cmdclass, 'external_key_authority_contract', do_external_key_authority_contract)
